src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,444 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implements utility methods for working with some complex BDX messages.
+ */
+
+#include <protocols/bdx/BDXMessageUtils.h>
+
+#include <support/BufBound.h>
+#include <support/BufferReader.h>
+#include <support/CodeUtils.h>
+
+#define VERSION_MASK 0x0F
+#define SENDER_DRIVE_MASK 0x10
+#define RECEIVER_DRIVE_MASK 0x20
+#define ASYNC_MASK 0x40
+#define CONTROL_MODE_MASK 0xF0
+
+#define DEFLEN_MASK 0x01
+#define START_OFFSET_MASK 0x02
+#define WIDERANGE_MASK 0x10
+
+using namespace chip;
+using namespace chip::BDX;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR TransferInit::Pack(System::PacketBuffer * aBuffer)
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rangeCtl = 0, ptcByte = 0;
+
+    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
+    if (aBuffer == NULL)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    BufBound bbuf(aBuffer->Start(), aBuffer->AvailableDataLength());
+
+    ptcByte |= mSupportedVersions & VERSION_MASK;
+    if (mSupportsSenderDrive)
+        ptcByte |= SENDER_DRIVE_MASK;
+    if (mSupportsReceiverDrive)
+        ptcByte |= RECEIVER_DRIVE_MASK;
+    if (mSupportsAsync)
+        ptcByte |= ASYNC_MASK;
+
+    bbuf.Put(ptcByte);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    if (mDefLen)
+        rangeCtl |= DEFLEN_MASK;
+    if (mStartOffset)
+        rangeCtl |= START_OFFSET_MASK;
+    if (mWideRange)
+        rangeCtl |= WIDERANGE_MASK;
+
+    bbuf.Put(rangeCtl);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    bbuf.PutLE16(mMaxBlockSize);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    if (mStartOffset > 0)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mStartOffset);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+        else
+        {
+            bbuf.PutLE32((uint32_t) mStartOffset);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+    }
+
+    if (mDefLen)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mMaxLength);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+        else
+        {
+            bbuf.PutLE32((uint32_t) mMaxLength);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+    }
+
+    VerifyOrExit(mFileDesignator && mFileDesLength > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    bbuf.PutLE16(mFileDesLength);
+    bbuf.Put(mFileDesignator, (size_t) mFileDesLength);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Metadata is optional
+    if (mMetadata && mMetadataLength > 0)
+    {
+        bbuf.Put(mMetadata, (size_t) mMetadataLength);
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
```

Comment:
this function will be a bunch smaller without all the VerifyOrExit() copy-pasta...

you don't have to check bbuf.Fit() until the end of the function, as bbuf will never overflow (that's it's whole job :-) )

Reply:
That makes sense! I'll remove the extra ones then

Reply:
So, what happens to the data if it doesn't fit? If it's truncated, the receiver may blow-up. It's likely a good idea to error-out on a client trying to include too much metadata.